src/dragonball/src/vcpu/vcpu_manager.rs

```diff
@@ -0,0 +1,1005 @@
+    #[error("unexpected internal error")]
+    InternalError,
+
+    /// vCPU manager is not initialized
+    #[error("vcpu manager is not initialized")]
+    VcpuManagerNotInitialized,
+
+    /// Expected vcpu exceed max count
+    #[error("expected vcpu exceed max count")]
+    ExpectedVcpuExceedMax,
+
+    /// vCPU not found
+    #[error("vcpu not found {0}")]
+    VcpuNotFound(u8),
+
+    /// Cannot recv vCPU thread tid
+    #[error("cannot get vCPU thread id")]
+    VcpuGettid,
+
+    /// vCPU pause failed.
+    #[error("failure while pausing vCPU thread")]
+    VcpuPause,
+
+    /// vCPU resume failed.
+    #[error("failure while resuming vCPU thread")]
+    VcpuResume,
+
+    /// vCPU save failed.
+    #[error("failure while save vCPU state")]
+    VcpuSave,
+
+    /// Vcpu is in unexpected state.
+    #[error("Vcpu is in unexpected state")]
+    UnexpectedVcpuResponse,
+
+    /// Vcpu not create
+    #[error("Vcpu is not create")]
+    VcpuNotCreate,
+
+    /// The number of max_vcpu reached kvm's limitation
+    #[error("the number of max_vcpu reached kvm's limitation")]
+    MaxVcpuLimitation,
+
+    /// Revalidate vcpu IoManager cache failed.
+    #[error("failure while revalidating vcpu IoManager cache")]
+    VcpuRevalidateCache,
+
+    /// Response channel error
+    #[error("Response channel error: {0}")]
+    VcpuResponseChannel(RecvError),
+
+    /// Vcpu response timeout
+    #[error("Vcpu response timeout: {0}")]
+    VcpuResponseTimeout(RecvTimeoutError),
+
+    /// Cannot build seccomp filters.
+    #[error("failure while configuring seccomp filters: {0}")]
+    SeccompFilters(#[source] seccompiler::Error),
+
+    /// Cannot send event to vCPU.
+    #[error("failure while sending message to vCPU thread: {0}")]
+    VcpuEvent(#[source] VcpuError),
+
+    /// vCPU Error
+    #[error("vcpu internal error: {0}")]
+    Vcpu(#[source] VcpuError),
+
+    #[cfg(feature = "hotplug")]
+    /// vCPU resize error
+    #[error("resize vcpu error: {0}")]
+    VcpuResize(#[source] VcpuResizeError),
+
+    /// Kvm Ioctl Error
+    #[error("failure in issuing KVM ioctl command: {0}")]
+    Kvm(#[source] kvm_ioctls::Error),
+}
+
+#[cfg(feature = "hotplug")]
+/// Errror associated with resize instance
+#[derive(Debug, thiserror::Error)]
+pub enum VcpuResizeError {
+    /// vcpu is in hotplug process
+    #[error("vcpu is in hotplug process")]
+    VcpuIsHotplugging,
+
+    /// Cannot update the configuration of the microvm pre boot.
+    #[error("resize vcpu operation is not allowed after boot")]
+    UpdateNotAllowedPostBoot,
+
+    /// Expected vcpu exceed max count
+    #[error("expected vcpu exceed max count")]
+    ExpectedVcpuExceedMax,
+
+    /// vcpu 0 can't be removed
+    #[error("vcpu 0 can't be removed")]
+    Vcpu0CanNotBeRemoved,
+
+    /// Lack removable vcpu
+    #[error("Removable vcpu not enough, removable vcpu num: {0}, number to remove: {1}, present vcpu count {2}")]
+    LackRemovableVcpus(u16, u16, u16),
```

Comment:
Related to my comment about `VcpuConfig`, this seems to suggest the cpu counts should be `u16`?